### PR TITLE
Add Molien series for the non-modular case

### PR DESCRIPTION
**Pull request by thofma**
I added the polynomial ring as an optional first argument and I also renamed `invariant_basis` to `basis` (because it is called on the invariant ring, which makes the `invariant_` part redundant).

Thanks for @ThomasBreuer for the GAP code.

P.S.: @joschmitt Why can Singular not handle non-prime finite fields?

### CODE REVIEW

**Comment by tthsqe12** [comment]
I don't like seeing the denominators non-factored. Do you want a FactoredField in AA/Oscar?

**Comment by thofma** [comment]
I don't want to use this PR to make these changes. If we go this route I suggest a dedicated type for Hilbert series, which can store and print the polynomials whatever way we want. Also we can add the functionality to get the coefficients of the Laurent series expansion easily.

File: experimental/InvariantTheory/invariant_rings.jl [hunk 1]
```diff
@@ -423,3 +359,107 @@ function heisenberg_group(n::Int)
...
+  R = polynomial_ring(IR)
+  K = coefficient_ring(IR)
+  # Need an extra coercion here while waiting for https://github.com/Nemocas/AbstractAlgebra.jl/pull/1009
+  #return to_univariate(S, R(mol[1, 1]))//to_univariate(S, R(mol[1, 2]))
+  return S(to_univariate(S, R(mol[1, 1])))//S(to_univariate(S, R(mol[1, 2])))
+end
+
+function _molien_series_via_singular(S::PolyRing, IR::InvRing{T}) where {T <: Union{Nemo.GaloisField, Nemo.GaloisFmpzField}}
+  @assert !ismodular(IR)
+  mol = molien_series_via_singular(IR)
```
Comment (fingolfin) [inline experimental/InvariantTheory/invariant_rings.jl]: Should this be this? If not, why not?
```suggestion
  mol = __molien_series_via_singular(IR)
```

File: experimental/InvariantTheory/invariant_rings.jl [hunk 2]
```diff
@@ -385,7 +385,7 @@ end
 
 function _molien_series_via_singular(S::PolyRing, IR::InvRing{T}) where {T <: Union{Nemo.GaloisField, Nemo.GaloisFmpzField}}
   @assert !ismodular(IR)
-  mol = molien_series_via_singular(IR)
+  mol = __molien_series_via_singular(IR)
 
   # TODO: Write a conversion from Singular number fields to Nemo ones
```
Comment (tthsqe12) [inline experimental/InvariantTheory/invariant_rings.jl:586]: This should probably go under the `oscar_ring` function?

Comment (thofma) [reply experimental/InvariantTheory/invariant_rings.jl:586]: This is something from @joschmitt's previous PR.

Comment (joschmitt) [reply experimental/InvariantTheory/invariant_rings.jl:586]: Yes, it would be nice to have `oscar_ring` for converting a Singular number field to Nemo.

**Review by fingolfin — APPROVED** [review]

File: experimental/InvariantTheory/invariant_rings.jl [hunk 3]
```diff
@@ -423,3 +359,107 @@ function heisenberg_group(n::Int)
...
+julia> G = MatrixGroup(3, K, [M1, M2]);
+
+julia> IR = invariant_ring(G);
+
+julia> molien_series(G)
+(-t^6 + t^3 - 1)//(t^9 - 3*t^6 + 3*t^3 - 1)
+```
+"""
+function molien_series(I::InvRing)
+  S, t = PolynomialRing(coefficient_ring(I), "t", cached = false)
```
Comment (joschmitt) [inline experimental/InvariantTheory/invariant_rings.jl]: I'm confused. Doesn't one have to use a different (characteristic 0) coefficient ring for the Hilbert series in the characteristic p case?

Comment (fingolfin) [reply experimental/InvariantTheory/invariant_rings.jl]: Indeed, good catch!

Also, if we want to keep a factorized representation, the coefficient ring must be some a priori unknown number field (of course as a rational function, expanded, it is defined over Q). I guess all of these aspects need to be taken care of by the to-be-added HilbertSeries type.

Comment (thofma) [reply experimental/InvariantTheory/invariant_rings.jl]: Hm, oops, of course! I guess we should always use a polynomial ring over the rationals by default? If I understand it correctly, the Molien series will always be a rational function over the rationals?

Comment (thofma) [reply experimental/InvariantTheory/invariant_rings.jl]: I think the idea will be to compute the factorization from the rational function f/g with f, g in Q[t]. @tthsqe12 is looking into it. In the meantime, I will change it to always use a rational polynomial ring by default.

Comment (fingolfin) [reply experimental/InvariantTheory/invariant_rings.jl]: Actually, for most uses, I guess it's fine to do all over Q: If I am not mistaken, the Hilbert series always can be written as `f(t) / [ (1-t^{d_1} ) \cdots (1-t^{d_n}) ]` for suitable exponents d_1, ..., d_n. While one could factor this further over a suitable cylotomic field extension, I don't think this is necessary. Indeed, for many applications, just getting `f` and the tuple `(d_1, ..., d_n)` will be enough.

Comment (tthsqe12) [reply experimental/InvariantTheory/invariant_rings.jl]: @fingolfin Is there some canonical representation of the Molien series (EDIT: that keeps only terms of the form 1-x^a on the bottom)? I would like to at least be able to tell the order of growth of the coefficients just by looking at the denominator? i.e. 1/(1-x) -> bounded, 1/((1-x)*(1-x^2)) -> degree 1, ect

Comment (thofma) [reply experimental/InvariantTheory/invariant_rings.jl]: No, in general there is not. That's why this is all some magic way to print it nicely, but without mathematical justification.

Comment (tthsqe12) [reply experimental/InvariantTheory/invariant_rings.jl]: Here is another example
```
  1-x+x^2           1+x^3
------------ = --------------
(1-x)(1-x^3)   (1-x^2)(1-x^3)
```

Probably we would go with the first one, but the second one allows for an easy proof that the coefficients are non-negative.

Comment (joschmitt) [reply experimental/InvariantTheory/invariant_rings.jl]: Regarding the coefficient ring: Singular constructs a cyclotomic field in this situation, but I have the impression that for GAP QQ is enough.

Comment (tthsqe12) [reply experimental/InvariantTheory/invariant_rings.jl]: @thofma so you have absolutely no desire to see positive coefficients on top?

Comment (thofma) [reply experimental/InvariantTheory/invariant_rings.jl]: Maybe we should ask someone who would like to have those. @fingolfin Any wishes on how it should look like? Or maybe @wdecker?

Comment (wdecker) [reply experimental/InvariantTheory/invariant_rings.jl]: Let me recall: In the 19th century, people played with different representations of the Hilbert series in order to make guesses about where to find generators for the invariant ring. In the context of computing invariants of finite groups, the only reasonable factorization is that corresponding to the degrees of the primary invariants. So how about leaving the Molien series unfactored until the primary invariants are computed, and cache the meaningful factorization then.

Comment (tthsqe12) [reply experimental/InvariantTheory/invariant_rings.jl]: Thanks for the clarification. I was getting anxious over all of the different possible representations.

Comment (wdecker) [reply experimental/InvariantTheory/invariant_rings.jl]: @tthsqe12: If unfactored, the 1 in the denominator should be 1 and not -1. This fits with the eventual meaningful factorization which exists over ZZ.

Comment (wdecker) [reply experimental/InvariantTheory/invariant_rings.jl]: Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.

Comment (wdecker) [reply experimental/InvariantTheory/invariant_rings.jl]: Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.

Comment (wdecker) [reply experimental/InvariantTheory/invariant_rings.jl]: Sorry, I was to quick. Of course, there is the theorem of Hilbert-Serre and thus another meaningful way of representing the Hilbert series which reflects the degrees of the fundamental invariants in the denominator (which do not(!!!) depend on the choice of fundamental invariants). Often, however, we do not wish to compute the fundamental invariants. But if they are computed, we could  also cache this factorization. I will think about this more carefully.

**Comment by joschmitt** [comment]
> P.S.: @joschmitt Why can Singular not handle non-prime finite fields?

We already discussed this, but for everybody: I'm not aware of a theoretical reason, it's just not implemented.

**Comment by fingolfin** [comment]
Merge?

**Comment by wdecker** [comment]
@fingolfin Yes

**Review by joschmitt — APPROVED** [review]

**Comment by thofma** [comment]
No, I have to fix the doc build first. Will do this when I am back.

**Comment by fingolfin** [comment]
Take your time! Thanks!

**Comment by joschmitt** [comment]
So far the molien series does not get cached. I'm not sure at what stage one should do it since you let the user supply a polynomial ring, but at least if one calls `molien_series(::InvRing)` it should be cached in my opinion.

**Comment by thofma** [comment]
Yes. I won't do it in this PR.

**Pull request merged**